<?xml version="1.0"?>
<launch>
  <!-- === Args básicos de simulación === -->
  <arg name="world_name" default="racecar" />
  <arg name="gui"        default="true" />
  <arg name="racecar_version" default="racecar-v2" />

  <!-- === Args de percepción / frames === -->
  <arg name="scan_topic"  default="/scan"/>
  <arg name="laser_frame" default="laser"/>
  <arg name="out_frame"   default="odom"/>   <!-- marco geométrico de trabajo -->
  <arg name="resample_ds" default="0.25"/>   <!-- paso uniforme para centerline_sampled -->
  
  <!-- ==== (arriba, junto al resto de args) Parámetros para la pista de conos ==== -->
  <arg name="spawn_cones"   default="true"/>

  <arg name="cones_pattern" default="corridor"/>  <!-- corridor | arc | chicane | slalom -->
  <arg name="cones_W"       default="1.6"/>       <!-- ancho pista [m] -->
  <arg name="cones_DS"      default="2.0"/>       <!-- paso longitudinal [m] -->
  <arg name="cones_N"       default="8"/>         <!-- nº de pares -->
  <arg name="cones_X0"      default="3.0"/>       <!-- offset x inicial -->
  <arg name="cones_Y0"      default="0.0"/>
  <arg name="cones_YAW0"    default="0.0"/>       <!-- rad, para corridor/chicane -->

  <!-- específicos de cada patrón -->
  <arg name="cones_ARC_R"   default="12.0"/>      <!-- radio arco [m] -->
  <arg name="cones_SL_A"    default="0.7"/>       <!-- amplitud eslalon [m] -->
  <arg name="cones_SL_L"    default="8.0"/>       <!-- longitud de onda eslalon [m] -->
  <arg name="cones_CH_DY"   default="1.0"/>       <!-- desplazamiento lateral chicane [m] -->
  <arg name="cones_CH_M"    default="4"/>         <!-- nº de pares primer tramo chicane -->

  <arg name="cones_clear"   default="true"/>      <!-- borra conos previos -->
  <arg name="cones_prefix"  default="cone"/>      <!-- prefijo de modelos -->

  <!-- Simulación usa clock de Gazebo -->
  <param name="use_sim_time" value="true"/>

  <!-- === Gazebo + mundo === -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find racecar_sim_gazebo)/worlds/$(arg world_name).world"/>
    <arg name="gui" value="$(arg gui)"/>
  </include>
  <!-- ==== Generador de pista de conos (Gazebo) ==== -->
  <!--<group if="$(arg spawn_cones)">
    <include file="$(find racecar_sim_gazebo)/launch/spawn_cones.launch">
      <arg name="pattern"        value="$(arg cones_pattern)"/>
      <arg name="W"              value="$(arg cones_W)"/>
      <arg name="DS"             value="$(arg cones_DS)"/>
      <arg name="N"              value="$(arg cones_N)"/>
      <arg name="X0"             value="$(arg cones_X0)"/>
      <arg name="Y0"             value="$(arg cones_Y0)"/>
      <arg name="YAW0"           value="$(arg cones_YAW0)"/>
      <arg name="ARC_R"          value="$(arg cones_ARC_R)"/>
      <arg name="SL_A"           value="$(arg cones_SL_A)"/>
      <arg name="SL_L"           value="$(arg cones_SL_L)"/>
      <arg name="CH_DY"          value="$(arg cones_CH_DY)"/>
      <arg name="CH_M"           value="$(arg cones_CH_M)"/>
      <arg name="clear_existing" value="$(arg cones_clear)"/>
      <arg name="prefix"         value="$(arg cones_prefix)"/>
    </include>
  </group>-->
   
  <node pkg="racecar_sim_gazebo" type="spawn_cones_zigzag.py" name="spawn_cones_zigzag" output="screen">
  <!-- Ancho de pista -->
  <param name="W" value="3.0" />
  <!-- Separación aprox. entre parejas de conos (y entre conos del mismo lado) -->
  <param name="DS" value="1.0" />
  <!-- Número de semicircunferencias -->
  <param name="N_arcs" value="8" />
  <!-- Radio de cada semicircunferencia -->
  <param name="R_arc" value="4.0" />
  <!-- Pose inicial de la traza media -->
  <param name="X0" value="0.0" />
  <param name="Y0" value="0.0" />
  <param name="YAW0" value="0.0" />
  <!-- Limpia conos anteriores con el mismo prefijo -->
  <param name="clear_existing" value="true" />
  <param name="prefix" value="cone_zigzag" />
  </node>


  <!-- === Modelo URDF (xacro) === -->
  <param name="robot_description"
         command="$(find xacro)/xacro '$(find racecar_description)/urdf/racecar.xacro'"/>

  <!-- === Spawn del robot en Gazebo === -->
  <node name="racecar_spawn" pkg="gazebo_ros" type="spawn_model" output="screen"
        args="-urdf -param robot_description -model racecar -z 0.05" />

  <!-- === Control (ros_control) === -->
  <include file="$(find racecar_sim_control)/launch/racecar_control.launch" ns="/"/>

  <!-- === MUXs (teleop / ackermann mux) === -->
  <include file="$(find racecar_ackermann_mux)/launch/mux.launch" ns="vesc" />

  <!-- === Relay de odometría (como en tu setup) === -->
  <node name="better_odom" pkg="topic_tools" type="relay"
        args="/vesc/odom /pf/pose/odom" output="screen"/>

  <!-- === TF estático map->odom identidad (opcional pero útil para RViz) === -->
  <node pkg="tf" type="static_transform_publisher" name="map_to_odom"
        args="0 0 0 0 0 0 map odom 100"/>

  <!-- === Detección desde Hokuyo (/scan) → /cones/raw === -->
  <include file="$(find racecar_cone_detection_lidar)/launch/scan_to_cones.launch">
    <!-- Si quieres afinar, puedes pasar más args (eps0, eps_k, w_min, w_max, etc.) -->
  </include>
  

  <!-- === Geometría de pista: L–R pairing, midpoints, centerline, viz === -->
  <include file="$(find racecar_track_geometry)/launch/track_geometry.launch"/>

  <!-- (Opcional) Forzar ds del resampler por parámetro privado -->
  <param name="centerline_resampler/resample_ds" value="$(arg resample_ds)"/>

  <!-- === RViz opcional (puedes cargar tu config) ===
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find racecar_description)/rviz/racecar.rviz" />
  -->
</launch>
